### PR TITLE
move user and peer accounts in separate packages

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
## Reasoning behind the pull request
- move user and peer account in a separate package

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: common/interface.go [hunk 1]
```diff
@@ -424,3 +426,87 @@ type StateSyncNotifierSubscriber interface {
 	MissingDataTrieNodeFound(hash []byte)
 	IsInterfaceNil() bool
 }
+
+// PeerAccountHandler models a peer state account, which can journalize a normal account's data
+//  with some extra features like signing statistics or rating information
+type PeerAccountHandler interface {
```
Comment (iulianpascalau) [inline common/interface.go:432]: these new definitions should be kept inside the `state` package

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: api/groups/validatorGroup_test.go [hunk 2]
```diff
@@ -34,8 +34,8 @@ func TestNewValidatorGroup(t *testing.T) {
 }
 
 type ValidatorStatisticsResponse struct {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply api/groups/validatorGroup_test.go:36]: Added

File: common/interface.go [hunk 3]
```diff
@@ -2,6 +2,8 @@ package common
 
 import (
 	"context"
+	vmcommon "github.com/multiversx/mx-chain-vm-common-go"
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply common/interface.go:5]: Done.

File: state/factory/accountCreator.go [hunk 4]
```diff
@@ -2,18 +2,32 @@ package factory
...
 
+// ArgsAccountCreator holds the arguments needed to create a new account creator
+type ArgsAccountCreator struct {
+	Hasher              hashing.Hasher
+	Marshaller          marshal.Marshalizer
+	EnableEpochsHandler common.EnableEpochsHandler
+}
+
 // AccountCreator has method to create a new account
 type AccountCreator struct {
```
Comment (ssd04) [inline state/factory/accountCreator.go:23]: can we unexport this struct ?

File: epochStart/metachain/baseRewards_test.go [hunk 5]
```diff
@@ -18,13 +18,12 @@ import (
 	"github.com/multiversx/mx-chain-go/epochStart/mock"
 	"github.com/multiversx/mx-chain-go/process"
 	"github.com/multiversx/mx-chain-go/sharding"
-	"github.com/multiversx/mx-chain-go/state"
+	"github.com/multiversx/mx-chain-go/state/accounts"
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply epochStart/metachain/baseRewards_test.go:21]: Done.

File: state/accounts/userAccount.go [hunk 6]
```diff
@@ -1,58 +1,46 @@
...
 
-// ArgsAccountCreation holds the arguments needed to create a new instance of userAccount
-type ArgsAccountCreation struct {
-	Hasher              hashing.Hasher
-	Marshaller          marshal.Marshalizer
-	EnableEpochsHandler common.EnableEpochsHandler
-}
-
-// NewUserAccount creates a new instance of userAccount
+// NewUserAccount creates a new instance of accounts
```
Comment (ssd04) [inline state/accounts/userAccount.go:30]: new instance or user account?

File: testscommon/trie/trieLeafParserStub.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package trie
+
+import "github.com/multiversx/mx-chain-core-go/core"
+
+// generate stub code for TrieLeafParser interface
+
```
Comment (ssd04) [inline testscommon/trie/trieLeafParserStub.go:6]: remove this comment?

File: state/accountsDB_test.go [hunk 8]
```diff
@@ -2447,29 +2452,113 @@ func TestAccountsDB_Close(t *testing.T) {
...
+			},
+		}
+		adb, _ := state.NewAccountsDB(args)
+
+		acc, err := adb.GetAccountFromBytes([]byte{1}, []byte{})
+		assert.Nil(t, acc)
+		assert.Equal(t, expectedErr, err)
+	})
+
+	t.Run("unmarshall err", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/accountsDB_test.go:2485]: Done.

File: common/interface.go [hunk 9]
```diff
@@ -424,3 +426,87 @@ type StateSyncNotifierSubscriber interface {
...
+	GetOwnerAddress() []byte
+	SetUserName(userName []byte)
+	GetUserName() []byte
+	IsGuarded() bool
+	GetAllLeaves(leavesChannels *TrieIteratorChannels, ctx context.Context) error
+	vmcommon.AccountHandler
+}
+
+// DataTrieTracker models what how to manipulate data held by a SC account
+type DataTrieTracker interface {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply common/interface.go:498]: Done.

File: state/accounts/userAccount.go (hunk 6 above)
Comment (BeniaminDrasovean) [reply state/accounts/userAccount.go:30]: Done.

File: testscommon/trie/trieLeafParserStub.go (hunk 7 above)
Comment (BeniaminDrasovean) [reply testscommon/trie/trieLeafParserStub.go:6]: Done.

File: integrationTests/testNetwork.go [hunk 10]
```diff
@@ -330,7 +330,7 @@ func (net *TestNetwork) NewAddress(creator *TestWalletAccount) Address {
 	return address
 }
 
-// GetAccountHandler retrieves the `state.UserAccountHandler` instance for the
+// GetAccountHandler retrieves the `common.UserAccountHandler` instance for the
```
Comment (iulianpascalau) [inline integrationTests/testNetwork.go:333]: revert to `state.UserAccountHandler` ?

**Review by iulianpascalau — DISMISSED** [review]

File: state/trackableDataTrie.go [hunk 11]
```diff
@@ -48,7 +47,7 @@ func NewTrackableDataTrie(
 	}
 
 	return &trackableDataTrie{
-		tr:                  tr,
+		tr:                  nil,
```
Comment (ssd04) [inline state/trackableDataTrie.go:50]: why removing trie from constructor? we'll rely on setting it later with SetDataTrie?

File: trie/trieStorageManager.go [hunk 12]
```diff
@@ -580,6 +580,8 @@ func (tsm *trieStorageManager) RemoveFromCheckpointHashesHolder(hash []byte) {
 	tsm.storageOperationMutex.Lock()
 	defer tsm.storageOperationMutex.Unlock()
 
+	log.Trace("trie storage manager: remove", "hash", hash)
```
Comment (ssd04) [inline trie/trieStorageManager.go:583]: update log message to include RemoveFromCheckpointHashesHolder or something similar?

File: update/genesis/base.go [hunk 13]
```diff
@@ -101,14 +103,19 @@ func NewEmptyAccount(
...
+			return nil, err
 		}
-		return state.NewUserAccount(address, argsAccCreation)
+
+		dtlp, err := parsers.NewDataTrieLeafParser(address, marshaller, enableEpochsHandler)
+		if err != nil {
+			return nil, err
+		}
+
+		return accounts.NewUserAccount(address, dtt, dtlp)
```
Comment (ssd04) [inline update/genesis/base.go:116]: use CreateAccount from accountCreator here also?

File: state/accountsDB.go [hunk 14]
```diff
@@ -1060,20 +1060,30 @@ func (adb *AccountsDB) RecreateAllTries(rootHash []byte) (map[string]common.Trie
 	}
 
 	for leaf := range leavesChannels.LeavesChan {
-		account := &accounts.UserAccountData{}
+		account, err := adb.accountFactory.CreateAccount(leaf.Key())
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/accountsDB.go:1063]: Refactored.

File: integrationTests/testNetwork.go (hunk 10 above)
Comment (BeniaminDrasovean) [reply integrationTests/testNetwork.go:333]: Done.

File: state/trackableDataTrie.go (hunk 11 above)
Comment (BeniaminDrasovean) [reply state/trackableDataTrie.go:50]: When a new account is created, it does not have a data trie by default, so the trie will be nil at first. The data trie is created only when some data is added to the account.

File: trie/trieStorageManager.go (hunk 12 above)
Comment (BeniaminDrasovean) [reply trie/trieStorageManager.go:583]: Done.

File: update/genesis/base.go (hunk 13 above)
Comment (BeniaminDrasovean) [reply update/genesis/base.go]: Refactored.

**Review by iulianpascalau — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Pull request merged**